libnopegl/src/ngpu/opengl/cmd_buffer_gl.c

```diff
@@ -0,0 +1,217 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "log.h"
+
+#include "bindgroup_gl.h"
+#include "buffer_gl.h"
+#include "cmd_buffer_gl.h"
+#include "ctx_gl.h"
+#include "fence_gl.h"
+#include "pipeline_gl.h"
+#include "rendertarget_gl.h"
+
+#include "ngpu/ctx.h"
+#include "utils/memory.h"
+#include "utils/refcount.h"
+
+struct ngpu_cmd_buffer_gl {
+    struct ngpu_ctx *gpu_ctx;
+    int type;
+    struct ngpu_fence_gl *fence;
+    struct darray cmds; // array of cmd_gl
+    struct darray refs; // array of ngli_rc pointers
+};
+
+struct ngpu_cmd_buffer_gl *ngpu_cmd_buffer_gl_create(struct ngpu_ctx *gpu_ctx)
+{
+    struct ngpu_cmd_buffer_gl *s = ngli_calloc(1, sizeof(*s));
+    if (!s)
+        return NULL;
+    s->gpu_ctx = gpu_ctx;
+    return s;
+}
+
+static void unref_rc(void *user_arg, void *data)
+{
+    struct ngli_rc **rcp = data;
+    NGLI_RC_UNREFP(rcp);
+}
+
+void ngpu_cmd_buffer_gl_freep(struct ngpu_cmd_buffer_gl **sp)
+{
+    struct ngpu_cmd_buffer_gl *s = *sp;
+    if (!s)
+        return;
+
+    ngpu_cmd_buffer_gl_wait(s);
+
+    ngli_darray_reset(&s->refs);
+    ngli_darray_reset(&s->cmds);
+
+    ngli_freep(sp);
+}
+
+int ngpu_cmd_buffer_gl_init(struct ngpu_cmd_buffer_gl *s, int type)
+{
+    s->type = type;
+
+    ngli_darray_init(&s->cmds, sizeof(struct ngpu_cmd_gl), 0);
+    ngli_darray_init(&s->refs, sizeof(struct ngli_rc *), 0);
+    ngli_darray_set_free_func(&s->refs, unref_rc, NULL);
+
+    return 0;
+}
+
+int ngpu_cmd_buffer_gl_ref(struct ngpu_cmd_buffer_gl *s, struct ngli_rc *rc)
+{
+    if (!ngli_darray_push(&s->refs, &rc))
+        return NGL_ERROR_MEMORY;
+
+    NGLI_RC_REF(rc);
+
+    return 0;
+}
+
+int ngpu_cmd_buffer_gl_begin(struct ngpu_cmd_buffer_gl *s)
+{
+    ngli_darray_clear(&s->refs);
+    ngli_darray_clear(&s->cmds);
+
+    return 0;
+}
+
+int ngpu_cmd_buffer_gl_push(struct ngpu_cmd_buffer_gl *s, const struct ngpu_cmd_gl *cmd)
+{
+    if (!ngli_darray_push(&s->cmds, cmd))
+        return NGL_ERROR_MEMORY;
+
+    return 0;
+}
+
+int ngpu_cmd_buffer_gl_submit(struct ngpu_cmd_buffer_gl *s)
+{
+    struct ngpu_ctx *gpu_ctx = s->gpu_ctx;
+    struct ngpu_ctx_gl *gpu_ctx_gl = (struct ngpu_ctx_gl *)gpu_ctx;
+    struct glcontext *gl = gpu_ctx_gl->glcontext;
+
+    struct ngpu_rendertarget *cur_rendertarget = NULL;
+    struct ngpu_pipeline *cur_pipeline = NULL;
+
+    struct ngpu_cmd_gl *cmds = ngli_darray_data(&s->cmds);
+    for (size_t i = 0; i < ngli_darray_count(&s->cmds); i++) {
+        struct ngpu_cmd_gl *cmd = &cmds[i];
+
+        switch (cmd->type) {
+        case NGPU_CMD_TYPE_GL_SET_INDEX_BUFFER:
```

Comment:
style: either add `{ }` here or remove them where unnecessary (for example viewport or scissor)

Reply:
Added {}.